### PR TITLE
DEPR: move NumericIndex._convert_tolerance to Index

### PR DESCRIPTION
**Pull request by topper-123**
Moves`_convert_tolerance` from `NumericIndex`to `Index` in preparation to remove `NumericIndex` and include numpy int/uint/float64 in the base `Index`.

xref #42717.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -3761,6 +3761,17 @@ def _convert_tolerance(self, tolerance, target: np.ndarray | Index) -> np.ndarra
         tolerance = np.asarray(tolerance)
         if target.size != tolerance.size and tolerance.size > 1:
             raise ValueError("list-like tolerance size must match target index size")
+        elif is_numeric_dtype(self) and not np.issubdtype(tolerance.dtype, np.number):
```
[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/indexes/base.py:3764]: The difference is just the error message, e.g.:

```python
>>> import pandas as pd
>>> index = pd.Index([5, 2, 1], dtype=pd.Float64Dtype)
>>> index.get_indexer([2, 5, 0], method="ffill", tolerance="s")
TypeError: '<=' not supported between instances of 'numpy.ndarray' and 'numpy.ndarray'  # main
ValueError: tolerance argument for Index must be numeric if it is a scalar: array('s', dtype='<U1')  # this PR
```
The new error message is the same as for numpy numeric dtypes which I think makes sense to have the same error type.

As far as I see, we don't have tests for Float32/64 indexes, so adding tests for the new error only is a bit difficult...

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3783]: ok thanks for taking a look

[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/indexes/base.py:3783]: I could add tests for the numeric extension dtypes, e.g. modeling them after the existing numpy numeric tests?

(I'd like to do that after getting `NumericIndex` merged into `Index`, so the current issue can be finished up first though).

[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:3783]: yah not worth futzing with for now

**Comment by topper-123** [comment]
Ping.

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -3780,6 +3780,17 @@ def _convert_tolerance(self, tolerance, target: np.ndarray | Index) -> np.ndarra
...
+            if tolerance.ndim > 0:
+                raise ValueError(
+                    f"tolerance argument for {type(self).__name__} must contain "
+                    "numeric elements if it is list type"
+                )
+
+            raise ValueError(
+                f"tolerance argument for {type(self).__name__} must be numeric "
```
[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/indexes/base.py:3791]: Yeah,  how that `Index` can handle a lot of different dtypes (numeric and non-numeric) that makes sense. I'll update.

[the remark this answers is not in the document]
Comment (topper-123) [reply pandas/core/indexes/base.py]: Error messages updated.

**Comment by topper-123** [comment]
This has also been updated :-)

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jbrockmendel** [comment]
thanks @topper-123

**Pull request merged**